File: testing/meraxis_test_trees.py
#!/usr/bin/env python

"""Doc-string here..."""

import numpy as np
import matplotlib.pyplot as plt

__author__ = "Simon Mutch"
__date__   = "12 Apr 2013"

catalog_desc = [
    ("id_MBP", 'q'),
    ("M_vir", 'f8'),
    ("n_particles", 'i4'),
    ("position_COM", ('f4', 3)),
    ("position_MBP", ('f4', 3)),
    ("velocity_COM", ('f4', 3)),
    ("velocity_MBP", ('f4', 3)),
    ("R_vir", 'f4'),
    ("R_halo", 'f4'),
    ("R_max", 'f4'),
    ("V_max", 'f4'),
    ("sigma_v", 'f4'),
    ("spin", ('f4', 3),),
    ("q_triaxial", 'f4'),
    ("s_triaxial", 'f4'),
    ("shape_eigen_vectors", ('f4', (3,3))),
    ("padding", 'S8'),
]

group_desc = [
    ("id", 'i4'),
    ("tree_flags", 'i4'),
    ("desc_id", 'i4'),
    ("dummy", 'i4'),
    ("file_offset", 'i4'),
    ("file_index", 'i4'),
    ("n_subgroups", 'i4'),
]

subgroup_desc = [
    ("id", 'i4'),
    ("tree_flags", 'i4'),
    ("desc_id", 'i4'),
    ("dummy", 'i4'),
    ("file_offset", 'i4'),
    ("file_index", 'i4'),
]

trees_header_desc = [
  ("n_groups", 'i4'),
  ("n_subgroups", 'i4'),
  ("n_halos_max", 'i4'),
  ("n_trees_subgroup", 'i4'),
  ("n_trees_group", 'i4'),
]

# Create the trees files
header = np.zeros(1, dtype=trees_header_desc)
header['n_groups'] = 3
header['n_subgroups'] = 3
header['n_halos_max'] = 6
header['n_trees_subgroup'] = 3
header['n_trees_group'] = 3

group = np.zeros(3, dtype=group_desc)
group['id']=[0,1,2]
group['tree_flags']=[1,1,1]
group['desc_id']=[0,1,1]
group['dummy']=[0,0,0]
group['file_offset']=[1,1,1]
group['file_index']=[0,1,1]
group['n_subgroups']=[1,1,1]

subgroup = np.zeros(3, dtype=subgroup_desc)
subgroup['id']=[0,1,2]
subgroup['tree_flags']=[1,1,1]
subgroup['desc_id']=[0,1,2]
subgroup['dummy']=[0,0,0]
subgroup['file_offset']=[1,1,1]
subgroup['file_index']=[0,1,2]

with open('test_step001_scan001.trees_horizontal.0', "wb") as fout:
    for v in header:
        v.tofile(fout)
    for g, s in zip(group, subgroup):
        for v in g:
            v.tofile(fout)
        for v in s:
            v.tofile(fout)

header = np.zeros(1, dtype=trees_header_desc)
header['n_groups'] = 2
header['n_subgroups'] = 3
header['n_halos_max'] = 6
header['n_trees_subgroup'] = 3
header['n_trees_group'] = 2

group = np.zeros(2, dtype=group_desc)
group['id']=[0,1]
group['tree_flags']=[1,2]
group['desc_id']=[0,1]
group['dummy']=[0,0]
group['file_offset']=[1,1]
group['file_index']=[0,1]
group['n_subgroups']=[1,2]

subgroup = np.zeros(3, dtype=subgroup_desc)
subgroup['id']=[0,1,2]
subgroup['tree_flags']=[1,1,1]
subgroup['desc_id']=[0,1,1]
subgroup['dummy']=[0,0,0]
subgroup['file_offset']=[1,1,1]
subgroup['file_index']=[0,1,1]

with open('test_step001_scan001.trees_horizontal.1', "wb") as fout:
    for v in header:
        v.tofile(fout)
    for v in group[0]:
        v.tofile(fout)
    for v in subgroup[0]:
        v.tofile(fout)
    for v in group[1]:
        v.tofile(fout)
    for v in subgroup[1]:
        v.tofile(fout)
    for v in subgroup[2]:
        v.tofile(fout)

header = np.zeros(1, dtype=trees_header_desc)
header['n_groups'] = 2
header['n_subgroups'] = 2
header['n_halos_max'] = 6
header['n_trees_subgroup'] = 2
header['n_trees_group'] = 2

group = np.zeros(2, dtype=group_desc)
group['id']=[0,1]
group['tree_flags']=[1,1]
group['desc_id']=[0,1]
group['dummy']=[0,0]
group['file_offset']=[1,1]
group['file_index']=[0,1]
group['n_subgroups']=[1,1]

subgroup = np.zeros(2, dtype=subgroup_desc)
subgroup['id']=[0,1]
subgroup['tree_flags']=[1,2]
subgroup['desc_id']=[0,1]
subgroup['dummy']=[0,0]
subgroup['file_offset']=[1,1]
subgroup['file_index']=[0,1]

with open('test_step001_scan001.trees_horizontal.2', "wb") as fout:
    for v in header:
        v.tofile(fout)
    for g, s in zip(group, subgroup):
        for v in g:
            v.tofile(fout)
        for v in s:
            v.tofile(fout)



# Now deal with the catalog files

group = np.zeros(3, dtype=catalog_desc)
group["id_MBP"] = [1,500,1000]
group["M_vir"] = [5e9,5e9,5e9]
group["n_particles"] = [100,100,100]
group["R_vir"] = [2, 2, 2]
group["V_max"] = [100, 100, 100]
with open('subfind_000.catalog_group_properties', "wb") as fout:
    for g in group:
        g.tofile(fout)

subgroup = np.zeros(3, dtype=catalog_desc)
subgroup = group.copy()
with open('subfind_000.catalog_subgroup_properties', "wb") as fout:
    for s in subgroup:
        s.tofile(fout)

group = np.zeros(2, dtype=catalog_desc)
group["id_MBP"] = [1,500]
group["M_vir"] = [5e9,10e9]
group["n_particles"] = [100,200]
group["R_vir"] = [2, 3]
group["V_max"] = [100, 200]
with open('subfind_001.catalog_group_properties', "wb") as fout:
    for g in group:
        g.tofile(fout)

with open('subfind_001.catalog_subgroup_properties', "wb") as fout:
    for s in subgroup:
        s.tofile(fout)

with open('subfind_002.catalog_group_properties', "wb") as fout:
    for g in group:
        g.tofile(fout)

subgroup = np.zeros(3, dtype=catalog_desc)
subgroup["id_MBP"] = [1,500]
subgroup["M_vir"] = [5e9,10e9]
subgroup["n_particles"] = [100,200]
subgroup["R_vir"] = [2, 3]
subgroup["V_max"] = [100, 200]
with open('subfind_002.catalog_subgroup_properties', "wb") as fout:
    for s in subgroup:
        s.tofile(fout)

File: testing/meraxis_test_trees.c
#include <stdio.h>
#include <string.h>

#define STRLEN 256

//! This is the structure for a halo in the catalog files
struct catalog_halo_struct{
  long long id_MBP;                    //!< ID of most bound particle in structure
  double    M_vir;                     //!< Bryan & Norman (ApJ 495, 80, 1998) virial mass [M_sol/h]
  int       n_particles;               //!< Number of particles in the structure
  float     position_COM[3];           //!< Centre-of-mass position      [Mpc/h]
  float     position_MBP[3];           //!< Most bound particle position [Mpc/h]
  float     velocity_COM[3];           //!< Centre-of-mass velocity      [km/s]
  float     velocity_MBP[3];           //!< Most bound particle velocity [km/s]
  float     R_vir;                     //!< Virial radius [Mpc/h]
  float     R_halo;                    //!< Distance of last halo particle from MBP [Mpc/h]
  float     R_max;                     //!< Radius of maximum circular velocity     [Mpc/h]
  float     V_max;                     //!< Maximum circular velocity               [km/s]
  float     sigma_v;                   //!< Total 3D velocity dispersion            [km/s]
  float     spin[3];                   //!< Specific angular momentum vector        [Mpc/h*km/s]
  float     q_triaxial;                //!< Triaxial shape parameter q=b/a
  float     s_triaxial;                //!< Triaxial shape parameter s=c/a
  float     shape_eigen_vectors[3][3]; //!< Normalized triaxial shape eigenvectors
  char      padding[8];                //!< Alignment padding
};
typedef struct catalog_halo_struct catalog_halo_struct;

static void write_trees_header(
  FILE *fout,            
  int   n_groups,        
  int   n_subgroups,     
  int   n_halos_max,     
  int   n_trees_subgroup,
  int   n_trees_group)   
{
  fwrite(&n_groups         , 1, sizeof(int), fout);
  fwrite(&n_subgroups      , 1, sizeof(int), fout);
  fwrite(&n_halos_max      , 1, sizeof(int), fout);
  fwrite(&n_trees_subgroup , 1, sizeof(int), fout);
  fwrite(&n_trees_group    , 1, sizeof(int), fout);
}

static void write_catalogs_header(
    FILE *fout,
    int i_file,
    int N_files,
    int N_halos_file,
    int N_halos_total)
{
  fwrite(&i_file       , 1, sizeof(int), fout);
  fwrite(&N_files      , 1, sizeof(int), fout);
  fwrite(&N_halos_file , 1, sizeof(int), fout);
  fwrite(&N_halos_total, 1, sizeof(int), fout);
}

static void write_group(
    FILE *fout,
    int id,
    int tree_flags,
    int desc_id,
    int file_offset,
    int file_index,
    int n_subgroups)
{
  int dummy = 0;

  fwrite(&id         , 1, sizeof(int), fout);
  fwrite(&tree_flags , 1, sizeof(int), fout);
  fwrite(&desc_id    , 1, sizeof(int), fout);
  fwrite(&dummy      , 1, sizeof(int), fout);
  fwrite(&file_offset, 1, sizeof(int), fout);
  fwrite(&file_index , 1, sizeof(int), fout);
  fwrite(&n_subgroups, 1, sizeof(int), fout);
}

static void write_subgroup(
    FILE *fout,
    int id,
    int tree_flags,
    int desc_id,
    int file_offset,
    int file_index)
{
  int dummy = 0;

  fwrite(&id         , 1, sizeof(int), fout);
  fwrite(&tree_flags , 1, sizeof(int), fout);
  fwrite(&desc_id    , 1, sizeof(int), fout);
  fwrite(&dummy      , 1, sizeof(int), fout);
  fwrite(&file_offset, 1, sizeof(int), fout);
  fwrite(&file_index , 1, sizeof(int), fout);
}

static void write_catalog_entry(
  FILE      *fout,       
  long long  id_MBP,     
  double     Mvir,       
  int        n_particles,
  float      R_vir,      
  float      V_max)      
{
  catalog_halo_struct entry;

  entry.id_MBP                    = id_MBP;
  entry.M_vir                     = Mvir;
  entry.n_particles               = n_particles;
  entry.position_COM[0]           = 0;
  entry.position_COM[1]           = 1;
  entry.position_COM[2]           = 2;
  entry.position_MBP[0]           = 0;
  entry.position_MBP[1]           = 1;
  entry.position_MBP[2]           = 2;
  entry.velocity_COM[0]           = 0;
  entry.velocity_COM[1]           = 1;
  entry.velocity_COM[2]           = 2;
  entry.velocity_MBP[0]           = 0;
  entry.velocity_MBP[1]           = 1;
  entry.velocity_MBP[2]           = 2;
  entry.R_vir                     = R_vir;
  entry.R_halo                    = R_vir;
  entry.R_max                     = R_vir;
  entry.V_max                     = V_max;
  entry.sigma_v                   = 100.0;
  entry.spin[0]                   = 0;
  entry.spin[1]                   = 1;
  entry.spin[2]                   = 2;
  entry.q_triaxial                = 1;
  entry.s_triaxial                = 1;
  entry.shape_eigen_vectors[0][0] = 0;
  entry.shape_eigen_vectors[0][1] = 0;
  entry.shape_eigen_vectors[0][2] = 0;
  entry.shape_eigen_vectors[1][0] = 0;
  entry.shape_eigen_vectors[1][1] = 0;
  entry.shape_eigen_vectors[1][2] = 0;
  entry.shape_eigen_vectors[2][0] = 0;
  entry.shape_eigen_vectors[2][1] = 0;
  entry.shape_eigen_vectors[2][2] = 0;

  fwrite(&entry, 1, sizeof(catalog_halo_struct), fout);
}

int main(int argc, char const* argv[])
{

  int snapshot;
  char trees_fname_base[STRLEN] = "halos/test/trees/horizontal/trees/test_step001_scan001.trees_horizontal";
  char fname[STRLEN];
  FILE *fout;
  FILE *fout_groups;
  FILE *fout_subgroups;

  /*
   * Snapshot 0
   */
  snapshot = 0;

  // Start with the trees
  sprintf(fname, "%s_%d", trees_fname_base, snapshot);
  fout = fopen(fname, "wb");

  write_trees_header(fout, 3, 3, 6, 3, 3);
  write_group(fout   , 0, 1, 0, 1, 0  , 1);
  write_subgroup(fout, 0, 1, 0, 1, 0);
  write_group(fout   , 1, 1, 1, 1, 1  , 1);
  write_subgroup(fout, 1, 1, 1, 1, 1);
  write_group(fout   , 2, 1, 2, 1, 2  , 1);
  write_subgroup(fout, 2, 1, 2, 1, 2);
  fclose(fout);

  // now deal with the catalogs
  sprintf(fname, "halos/test/catalogs/subfind_%03d.catalog_groups_properties", snapshot);
  fout_groups = fopen(fname, "wb");
  sprintf(fname, "halos/test/catalogs/subfind_%03d.catalog_subgroups_properties", snapshot);
  fout_subgroups = fopen(fname, "wb");

  write_catalogs_header(fout_groups   , 0, 1, 3, 3);
  write_catalogs_header(fout_subgroups, 0, 1, 3, 3);
  write_catalog_entry(fout_groups   , 100 , 6e9, 600, 1, 80);
  write_catalog_entry(fout_subgroups, 100 , 5e9, 500, 1, 80);
  write_catalog_entry(fout_groups   , 1300, 8e9, 800, 1, 90);
  write_catalog_entry(fout_subgroups, 1300, 7e9, 700, 1, 90);
  write_catalog_entry(fout_groups   , 2200, 9e9, 900, 1, 95);
  write_catalog_entry(fout_subgroups, 2200, 8e9, 800, 1, 95);
  fclose(fout_groups);
  fclose(fout_subgroups);

  /*
   * Snapshot 1
   */
  snapshot = 1;

  // Start with the trees
  sprintf(fname, "%s_%d", trees_fname_base, snapshot);
  fout = fopen(fname, "wb");

  write_trees_header(fout, 2, 3  , 6, 3, 3);
  write_group(fout       , 0, 1  , 0, 1, 0  , 1);
  write_subgroup(fout    , 0, 1  , 0, 1, 0);
  write_group(fout       , 1, 1  , 1, 1, 1  , 2);
  write_subgroup(fout    , 1, 1|2, 1, 1, 1);
  write_subgroup(fout    , 2, 1|2, 1, 1, 1);
  fclose(fout);

  // now deal with the catalogs
  sprintf(fname, "halos/test/catalogs/subfind_%03d.catalog_groups_properties", snapshot);
  fout_groups = fopen(fname, "wb");
  sprintf(fname, "halos/test/catalogs/subfind_%03d.catalog_subgroups_properties", snapshot);
  fout_subgroups = fopen(fname, "wb");

  write_catalogs_header(fout_groups   , 0, 1, 2, 2);
  write_catalogs_header(fout_subgroups, 0, 1, 3, 3);
  write_catalog_entry(fout_groups   , 100 , 6e9   , 600 , 1, 80);
  write_catalog_entry(fout_subgroups, 100 , 5e9   , 500 , 1, 80);
  write_catalog_entry(fout_groups   , 1300, 1.7e10, 1800, 1, 120);
  write_catalog_entry(fout_subgroups, 1300, 8e9   , 800 , 1, 95);
  write_catalog_entry(fout_subgroups, 2200, 7e9   , 700 , 1, 90);
  fclose(fout_groups);
  fclose(fout_subgroups);

  /*
   * Snapshot 2
   */
  snapshot = 2;

  // Start with the trees
  sprintf(fname, "%s_%d", trees_fname_base, snapshot);
  fout = fopen(fname, "wb");

  write_trees_header(fout, 2, 2, 6, 3, 3);
  write_group(fout   , 0, 1, 0, 1, 0  , 1);
  write_subgroup(fout, 0, 1, 0, 1, 0);
  write_group(fout   , 1, 1, 1, 1, 1  , 2);
  write_subgroup(fout, 1, 1, 1, 1, 1);
  fclose(fout);

  // now deal with the catalogs
  sprintf(fname, "halos/test/catalogs/subfind_%03d.catalog_groups_properties", snapshot);
  fout_groups = fopen(fname, "wb");
  sprintf(fname, "halos/test/catalogs/subfind_%03d.catalog_subgroups_properties", snapshot);
  fout_subgroups = fopen(fname, "wb");

  write_catalogs_header(fout_groups   , 0, 1, 2, 2);
  write_catalogs_header(fout_subgroups, 0, 1, 2, 2);
  write_catalog_entry(fout_groups   , 100 , 6e9   , 600 , 1, 80);
  write_catalog_entry(fout_subgroups, 100 , 5e9   , 500 , 1, 80);
  write_catalog_entry(fout_groups   , 1300, 1.7e10, 1800, 1, 120);
  write_catalog_entry(fout_subgroups, 100 , 1.5e10, 1500, 1, 120);
  fclose(fout_groups);
  fclose(fout_subgroups);

  return 0;
}
